pkg/noun/trace.c

```diff
@@ -1120,3 +1123,106 @@ u3t_etch_meme(c3_l mod_l)
   }
 }
 
+/* u3t_sstack_init: initalize a root node on the spin stack 
+*/
+void
+u3t_sstack_init()
+{
+  c3_c shm_name[256];
+  snprintf(shm_name, sizeof(shm_name), SLOW_STACK_NAME, getppid());
+  c3_w shm_fd = shm_open(shm_name, O_CREAT | O_RDWR, 0666);
+  if ( -1 == shm_fd) {
+    perror("shm_open failed");
+    return;
+  }
+
+  if ( -1 == ftruncate(shm_fd, PSIZE)) {
+    perror("truncate failed");
+    return;
+  }
+
+  stk_u = mmap(NULL, PSIZE, PROT_READ | PROT_WRITE, MAP_SHARED, shm_fd, 0);
+  
+  if ( MAP_FAILED == stk_u ) {
+    perror("mmap failed");
+    return;
+  }
+
+  stk_u->off_w = 0;
+  stk_u->fow_w = 0;
+  u3t_sstack_push(c3__root);
+}
+
+/* u3t_sstack_open: initalize a root node on the spin stack 
+*/
+u3t_spin*
+u3t_sstack_open()
+{
+
+  //Setup spin stack
+  c3_c shm_name[256];
+  snprintf(shm_name, sizeof(shm_name), SLOW_STACK_NAME, getpid());
+  c3_w shm_fd = shm_open(shm_name, O_CREAT | O_RDWR, 0);
+  if ( -1 == shm_fd) {
+    perror("shm_open failed");
+    return NULL; 
+  }
+
+  u3t_spin* stk_u = mmap(NULL, PSIZE, 
+                      PROT_READ | PROT_WRITE, MAP_SHARED, shm_fd, 0);
+  
+  if ( MAP_FAILED == stk_u ) {
+    perror("mmap failed");
+    return NULL; 
+  }
+
+  return stk_u;
+}
+/* u3t_sstack_exit: shutdown the shared memory for thespin stack 
+*/
+void
+u3t_sstack_exit()
+{
+  munmap(stk_u, u3a_page);
+}
+
+/* u3t_sstack_push: push a noun on the spin stack.
+*/
+void
+u3t_sstack_push(u3_noun nam)
+{
```

Comment:
To handle cases where there's no shared memory stack (like running in the ivory pill):

```
if ( !stk_u ) {
  u3z(name);
  return;
}
```

Reply:
Done